src/Giraffe/Routing.fs

```diff
@@ -10,37 +10,89 @@ open FSharp.Control.Tasks.V2.ContextInsensitive
 open Giraffe.FormatExpressions
 
 // ---------------------------
-// Private sub route helper functions
+// Sub Routing Feature
 // ---------------------------
 
-[<RequireQualifiedAccess>]
-module SubRouting =
+[<Literal>]
+let private RouteKey = "giraffe_route"
+
+type ISubRoutingFeature =
+    abstract member GetResolvedPath   : unit   -> string option
+    abstract member SetResolvedPath   : string -> unit
+    abstract member ClearResolvedPath : unit   -> unit
+
+type SubRoutingFeature() =
+    let mutable resolvedPath = None
 
-    [<Literal>]
-    let private RouteKey = "giraffe_route"
+    interface ISubRoutingFeature with
+        member __.GetResolvedPath()             = resolvedPath
+        member __.SetResolvedPath (p : string)  = resolvedPath <- Some p
+        member __.ClearResolvedPath()           = resolvedPath <- None
 
-    let getSavedPartialPath (ctx : HttpContext) =
-        if ctx.Items.ContainsKey RouteKey
-        then ctx.Items.Item RouteKey |> string |> strOption
+type HttpContext with
+
+    [<Obsolete("This method has been deprecated. Please use 'GetResolvedPath' instead which is a more memory efficient implementation based on the HttpContext.Features API.")>]
+    member this.GetResolvedPathFromItems() =
+        if this.Items.ContainsKey RouteKey
+        then this.Items.Item RouteKey |> string |> strOption
         else None
 
-    let getNextPartOfPath (ctx : HttpContext) =
-        match getSavedPartialPath ctx with
-        | Some p when ctx.Request.Path.Value.Contains p -> ctx.Request.Path.Value.[p.Length..]
-        | _   -> ctx.Request.Path.Value
+    /// **Description**
+    ///
+    /// Returns a partially (or fully) resolved request path during routing decisions when working with the SubRoutingFeature.
+    ///
+    member this.GetResolvedPath() =
+        let subRoutingFeature = this.Features.Get<ISubRoutingFeature>()
+        match subRoutingFeature.GetResolvedPath() with
+        | Some p -> Some p
+        | None   -> this.GetResolvedPathFromItems()
+
+    /// **Description**
+    ///
+    /// Returns the next part of the request path, which hasn't been resolved by a routing handler yet.
+    ///
+    member this.GetNextPartOfPath() =
+        let requestPath = this.Request.Path.Value
+        match this.GetResolvedPath() with
+        // Should this be a StartsWith?
```

Comment:
Shouldn't `requestPath.Contains` actually be `requestPath.StartsWith` here?

Reply:
looking at the usage inside the match, I'd say yes for correctness. Looks like the contains was added to avoid an issue where the request path was changed completely to the error path by the error handling middleware: https://github.com/giraffe-fsharp/Giraffe/commit/a2538f45c4ffab4c5ee263b3658328d93f1cb766

Reply:
Coolio, I've changed it to StartsWith now.